src/integrationTest/sam.test.ts

```diff
@@ -0,0 +1,192 @@
+/*!
+ * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import * as assert from 'assert'
+import { mkdirpSync, readFileSync, removeSync } from 'fs-extra'
+import * as path from 'path'
+import * as vscode from 'vscode'
+import { SamLambdaRuntime } from '../../src/lambda/models/samLambdaRuntime'
+import { getSamCliContext } from '../../src/shared/sam/cli/samCliContext'
+import { runSamCliInit, SamCliInitArgs } from '../../src/shared/sam/cli/samCliInit'
+import { activateExtension, sleep, TIMEOUT } from './integrationTestsUtilities'
+
+const projectFolder = `${__dirname}`
+let projectSDK = ''
+let projectPath = ''
+let debuggerType = ''
+let documentUri: vscode.Uri
+
+const runtimes = [
+    { name: 'nodejs8.10', path: 'testProject/hello-world/app.js', debuggerType: 'node2' },
+    { name: 'nodejs10.x', path: 'testProject/hello-world/app.js', debuggerType: 'node2' },
+    { name: 'python2.7', path: 'testProject/hello_world/app.py', debuggerType: 'python' },
+    { name: 'python3.6', path: 'testProject/hello_world/app.py', debuggerType: 'python' },
+    { name: 'python3.7', path: 'testProject/hello_world/app.py', debuggerType: 'python' },
+    { name: 'dotnetcore2.1', path: 'testProject/src/HelloWorld/Function.cs', debuggerType: 'coreclr' }
+]
+
+async function openSamProject(): Promise<vscode.Uri> {
+    const documentPath = path.join(projectFolder, projectPath)
+    await vscode.workspace.openTextDocument(documentPath)
+
+    return vscode.Uri.file(documentPath)
+}
+
+function tryRemoveProjectFolder() {
+    try {
+        removeSync(path.join(projectFolder, 'testProject'))
+    } catch (e) {}
+}
+
+async function getCodeLenses(): Promise<vscode.CodeLens[]> {
+    let codeLenses: vscode.CodeLens[] | undefined
+    while (true) {
+        try {
+            // this works without a sleep locally, but not on CodeBuild
+            await sleep(200)
+            const codeLensesPromise: Thenable<vscode.CodeLens[] | undefined> = vscode.commands.executeCommand(
+                'vscode.executeCodeLensProvider',
+                documentUri
+            )
+            codeLenses = await codeLensesPromise
+            if (!codeLenses) {
+                continue
+            }
+            // omnisharp spits out some undefined code lenses for some reason, we filter them because they are
+            // not shown to the user and do not affect how our extension is working
+            codeLenses = codeLenses.filter(lens => lens !== undefined && lens.command !== undefined)
+            if (codeLenses.length === 3) {
```

Comment:
Can we assert that length is 3 by this point, or are there times where omnisharp returns a whole list of undefined code lenses? As an example, if the code was only returning 2 code lenses, we appear to enter an infinite loop.

Reply:
(ok I see you have a throttle in the next function. Still curious though)

Reply:
I think I was having issues on CodeBuild with it returning too fast for some reason, let me try that and I'll see if it it still passes on CodeBuild

Reply:
yep, .net returns codelenses before we want to on codebuild so we need to keep it the way it is